src/storage/page_buffer.c

```diff
@@ -7889,6 +7852,81 @@ pgbuf_claim_bcb_for_fix (THREAD_ENTRY * thread_p, const VPID * vpid, PAGE_FETCH_
   return bufptr;
 }
 
+int
+pgbuf_read_page_from_file_or_page_server (THREAD_ENTRY * thread_p, const VPID * vpid, void *io_page)
+{
+
+// Server type --- Storage Config Type --- Data Type --- Connection to PS --- Local Read --- Remote Read
+//     PS                Local            Perm & Temp          Never              Yes            No
+// -----------------------------------------------------------------------------------------------------
+//     TS                Local                Perm              Yes               Yes           Yes
+//                                                               No               Yes            No
+//                                            Temp             Yes/No             Yes            No
+//                       Remote               Perm              Yes                No           Yes
+//                                            Temp              Yes               Yes            No
+
+#if defined(SERVER_MODE)
+  bool is_tran_server = get_server_type () == SERVER_TYPE_TRANSACTION;
+  bool is_page_server = get_server_type () == SERVER_TYPE_PAGE;
+  bool is_temporary_page = pgbuf_is_temporary_volume (vpid->volid);
+  bool is_using_local_storage = !ats_Gl.uses_remote_storage ();
+  bool request_from_ps = is_tran_server && ats_Gl.is_page_server_connected () && !is_temporary_page;
+  bool read_from_local = is_page_server || (is_using_local_storage || is_temporary_page);
+  int error_code = NO_ERROR;
+
+  if (read_from_local)
+    {
+      error_code = pgbuf_read_page_from_file (thread_p, vpid, io_page);
+      if (error_code != NO_ERROR)
+	{
+	  ASSERT_ERROR ();
+	  return error_code;
+	}
+    }
+  if (request_from_ps)
+    {
+      pgbuf_request_data_page_from_page_server (vpid);
+      auto data_page = ats_Gl.get_data_page_broker ().wait_for_page (*vpid);
+      if (read_from_local)
+	{
+	  char buf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
+	  FILEIO_PAGE *received_io_page = (FILEIO_PAGE *) PTR_ALIGN (buf, MAX_ALIGNMENT);
+	  std::memcpy (received_io_page, data_page->c_str (), db_io_page_size ());
+	  assert (received_io_page->prv == static_cast < FILEIO_PAGE * >(io_page)->prv);
```

Comment:
- compare may used data_page directly, without copying it.
- disable indent when template is used.
```suggestion
	  char buf[IO_MAX_PAGE_SIZE + MAX_ALIGNMENT];
	  FILEIO_PAGE *received_io_page = (FILEIO_PAGE *) PTR_ALIGN (buf, MAX_ALIGNMENT);
	  std::memcpy (received_io_page, data_page->c_str (), db_io_page_size ());
	  // *INDENT-OFF*
	  assert (reinterpret_cast<FILEIO_PAGE *> (data_page->c_str ())->prv
	              == reinterpret_cast<FILEIO_PAGE *> (io_page)->prv);
	  // *INDENT-ON*
```